### PR TITLE
fix(workspace-tools): remove message prefix from `workspaces foreach`

### PR DESCRIPTION
**Pull request by qrohlf**
### What's the problem this PR addresses?

_problem statement taken from #2517_
> Context: I have a Typescript project with many interdependent workspaces, most of which are libraries for a handful of applications. In CI, I use the following command to build any particular app with all its workspace dependencies:
>
> ```
> yarn workspaces foreach -vpt --topological-dev -j 100 run build
> ```
>
> Each workspace's build script calls `tsc` with some flags.
> 
> I use GitHub Actions as my CI with the https://github.com/actions/setup-node action to configure my CI environment for building my project. That setup step by default includes a [tsc problem matcher](https://github.com/actions/setup-node/blob/master/.github/tsc.json) which uses regex to match the output of the Typescript compiler and annotate the code in a PR with the actual errors.
> 
> If my CI runs `yarn run build` on a workspace directly, the output is identical to the output of `tsc`, so the matcher works fine. However, since I use yarn workspaces foreach, the output gets prefixed with Yarn's log level codes, causing the problem matcher to not match the output correctly. In particular, it breaks on matching the file name, causing the annotation to end up in the wrong place.
> This could be solved by modifying the regex in the problem matcher, but I believe it's out of scope for a tsc problem matcher to account for all kinds of wrappers that might be mutating the compiler's error format.
>
> This could also be worked around by piping the output through another command that removes that prefix from the output, but my intuition is that'd be prone to failure. I imagine parsing the output of CLI tools in some way is a common enough use case that Yarn should have the flag to allow it without strange workarounds.

Closes https://github.com/yarnpkg/berry/issues/2517

### How did you fix it?

Based on @arcanis's [comment](https://github.com/yarnpkg/berry/issues/2517#issuecomment-789955613), I chose to remove the `➤ YN0000: ` prefix from the `yarn workspaces foreach` stream reporter output entirely, by adding an additional flag to the StreamReport options to allow suppressing the `➤ ` and message name.

This aligns with the existing behavior of the analogous single-workspace `yarn exec` command, which is one of the few Yarn Berry commands that does not add any kind of prefixing/formatting on top of its output.

Note that this is an unfamiliar codebase for me and I'm not sure if my code changes are the _cleanest_ way of attaining the desired result here. I'm pushing this up without updating any of the tests yet to get feedback on this approach - very open to alternative suggestions.

This only applies to the outer layer of execution; if the `enableMessageNames` config option is set to true, the sub-executions of `yarn workspaces foreach` will still include the `YNXXXX` output. 

### Checklist
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
In https://github.com/yarnpkg/berry/issues/2517#issuecomment-786004138 @arcanis wrote to replace the `StreamReport` with a `ThrowReport`, what made you go with modifying `StreamReport` instead? Any benefits of doing it this way?

File: packages/plugin-workspace-tools/sources/commands/foreach.ts [hunk 1]
```diff
@@ -218,9 +218,15 @@ export default class WorkspacesForeachCommand extends BaseCommand {
 
     let abortNextCommands = false;
 
+    // get the configuration again but override messageNames so that we don't double-nest message names
+    // in the foreach output stream if they're enabled.
+    const configWithoutMessageNames = await Configuration.find(this.context.cwd, this.context.plugins);
+    configWithoutMessageNames.values.set(`enableMessageNames`, false);
```
Comment (merceyz) [inline packages/plugin-workspace-tools/sources/commands/foreach.ts:224]: This doesn't seem ideal, the `StreamReport` could instead have a `includePrefix` option that disabled both the "caret" and the prefix.

Comment (qrohlf) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts:224]: Agree that this feels hacky and that feels cleaner, I'll go ahead and update the PR.

**Comment by qrohlf** [comment]
@merceyz alternatively your earlier suggestion to use a ThrowReport instead of a StreamReport seems like it could also work. 

Again, I'm not super familiar with this codebase so I'd be curious to know if there's a downside to swapping out StreamReport with ThrowReport within foreach that made you change your suggestion?

**Comment by merceyz** [comment]
> I'd be curious to know if there's a downside to swapping out StreamReport with ThrowReport within foreach that made you change your suggestion?

After checking the code a bit more I saw that it was easier to modify `StreamReport` so I essentially answered my own question.

File: .yarn/versions/4aa0961e.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+releases:
+  "@yarnpkg/core": patch
+  "@yarnpkg/plugin-workspace-tools": patch
```
Comment (qrohlf) [inline .yarn/versions/4aa0961e.yml:3]: I'm not sure if modifying the console output of `foreach` would be considered a breaking change?

Comment (merceyz) [reply .yarn/versions/4aa0961e.yml:3]: No, if the output is supposed to be used by tools (and thus be stable) we would have had a `--json` flag.

File: .yarn/versions/4aa0961e.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+releases:
+  "@yarnpkg/core": patch
```
Comment (merceyz) [inline .yarn/versions/4aa0961e.yml:2]:
```suggestion
  "@yarnpkg/core": minor
```
A new feature was added to the core.

**Review by merceyz — CHANGES_REQUESTED** [review]
`includePrefix` should apply to all the methods that print a prefix, not just `reportInfo`.

**Comment by qrohlf** [comment]
@merceyz PR updated - I went ahead and extracted the `formatUtils.pretty` and whitespace concatenation that was happening with the caret in all those `reportFoo` methods to a new private function as well, seemed like there was enough duplication happening in this file that it was worth it.

**Review by merceyz — APPROVED** [review]

**Comment by merceyz** [comment]
Thanks for the fix!

**Comment by lusarz** [comment]
When it will be released?

**Comment by merceyz** [comment]
It was released in https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F3.4.0

**Pull request merged**